### ISSUE TITLE
Challenge-32 - Nodemon error

### ISSUE THREAD

**Issue by Roger-Melo** [da2k/curso-javascript-ninja#2253]
Olá, professor.  

O nodemon está me mostrando o erro "Error: Can't set headers after they are sent." 
Você sabe o que pode estar causando isso? 

![image](https://user-images.githubusercontent.com/29297788/33386268-bd45b1bc-d511-11e7-9b27-f6360103e41a.png)

Código:  

![image](https://user-images.githubusercontent.com/29297788/33386355-efcbaf56-d511-11e7-92d6-8a04fca1d603.png)

![image](https://user-images.githubusercontent.com/29297788/33386456-21caa372-d512-11e7-8651-4ef3fc58fa5e.png)

@fdaciuk

**Comment by fdaciuk** [comment]
Oi @Roger-Melo! Provavelmente é porque você está dando dois `res.send` =)

Veja o seu `if (!hasCar)`.. se entrar nesse `if`, ele vai disparar no final um `res.send`, mas como você não tem `return` para parar a função e ignorar qualquer coisa que venha após o `if`, o próximo `res.send' também é disparado, gerando esse erro =)

Ficou claro?

**Comment by Roger-Melo** [comment]
Entendi, professor. Declarei o return e deu certo mesmo.  

Outra coisa é que, o código da aplicação está muito grande e com algumas duplicações, como nas `'tr` a serem inseridas na tabela através do formulário e através do ajax. O que vc faria para melhorar esse código gigante?  

```javascript
(function(win, doc, $) {
  'use strict';

  var app = (function appController() {
    return {
      init: function init() {
        this.companyInfo();
        this.registerNewCar();
        this.getRegistredCar();
        this.initEvents();
      },

      registerNewCar: function registerNewCar() {
        var ajax = new XMLHttpRequest();
        ajax.open('POST', 'http://localhost:3000/car');
        ajax.setRequestHeader('Content-Type', 'application/x-www-form-urlencoded');
        ajax.send('image=http://wikiapk.com/wp-content/uploads/2017/06/73765_featured.jpg&brandModel=audi&year=3000&plate=FUJ9865&color=transparent');

        console.log('Cadastrando carro...');

        ajax.addEventListener('readystatechange', function() {
          if(ajax.readyState === 4)
            console.log(JSON.parse(ajax.responseText).message);
        }, false);
      },

      getRegistredCar: function getRegistredCar() {
        var ajax = new XMLHttpRequest();
        ajax.open('GET', 'http://localhost:3000/car');
        ajax.send();

        ajax.addEventListener('readystatechange', function() {
          if(ajax.readyState === 4) {
            var cars = JSON.parse(ajax.responseText);
            cars.forEach(app.insertRegistredCarInTable);
          }
        }, false);
      },

      insertRegistredCarInTable: function insertRegistredCarInTable(car) {
        var $fragment = doc.createDocumentFragment();
        var $tr = doc.createElement('tr');
        $tr.appendChild(app.createRegistredCarImage(car.image));
        $tr.appendChild(app.createRegistredCarTd(car.brandModel));
        $tr.appendChild(app.createRegistredCarTd(car.year));
        $tr.appendChild(app.createRegistredCarTd(car.plate));
        $tr.appendChild(app.createRegistredCarTd(car.color));
        $tr.appendChild(app.createRemoveButton());
        $fragment.appendChild($tr);
        $('[data-js="car-table"]').get().appendChild($fragment);
      },

      createRegistredCarTd: function createRegistredCarTd(tdContent) {
        var $td = app.createTd();
        $td.innerHTML = tdContent;
        return $td;
      },

      createRegistredCarImage: function createRegistredCarImage(carImage) {
        var $tdImage = app.createTd();
        var $image = doc.createElement('img');
        $image.setAttribute('src', carImage);
        $tdImage.appendChild($image);
        return $tdImage;
      },

      companyInfo: function companyInfo() {
        var ajax = new XMLHttpRequest();
        ajax.open('GET', '/company.json', true);
        ajax.send();
        ajax.addEventListener('readystatechange', this.getCompanyInfo, false);
      },

      getCompanyInfo: function getCompanyInfo() {
        if(!app.isRequestOk.call(this))
          return;
        var data = JSON.parse(this.responseText);
        app.fillCompanyInfo(data.name, data.phone);
      },

      fillCompanyInfo: function fillCompanyInfo(dataName, dataPhone) {
        var $companyName = $('[data-js="company-name"]');
        var $companyPhone = $('[data-js="company-phone"]');
        $companyName.get().textContent = dataName;
        $companyPhone.get().textContent = dataPhone;
      },

      isRequestOk: function isRequestOk() {
        return this.readyState === 4 && this.status === 200;
      },

      initEvents: function initEvents() {
        $('[data-js="form-register"]').on('submit', this.handleSubmit);
      },

      handleSubmit: function handleSubmit(e) {
        e.preventDefault();
        app.insertCarInTable();
      },

      insertCarInTable: function insertCarInTable() {
        return $('[data-js="car-table"]').get().appendChild(app.createNewCar());
      },

      createNewCar: function createNewCar() {
        return this.appendToFragment(this.createCarTr());
      },

      appendToFragment: function appendToFragment(tr) {
        return doc.createDocumentFragment().appendChild(tr);
      },

      createCarTr: function createCarTr() {
        var $tr = doc.createElement('tr');
        $tr.appendChild(this.createCarImage('[data-js="image"]'));
        $tr.appendChild(this.createCarTd('[data-js="brand-model"]'));
        $tr.appendChild(this.createCarTd('[data-js="year"]'));
        $tr.appendChild(this.createCarTd('[data-js="plate"]'));
        $tr.appendChild(this.createCarTd('[data-js="color"]'));
        $tr.appendChild(this.createRemoveButton());
        return $tr;
      },

      createRemoveButton: function createRemoveButton() {
        var $tdButton = this.createTd();
        var $button = doc.createElement('button');
        $button.innerHTML = 'Remover';
        $button.classList.add('remove-btn');
        $button.addEventListener('click', this.handleClickRemoveButton, false);
        $tdButton.appendChild($button);
        return $tdButton;
      },

      handleClickRemoveButton: function handleClickRemoveButton(e) {
        e.preventDefault();
        var $carTable = $('[data-js="car-table"]').get();
        var $trClickedButton = this.parentNode.parentNode;
        $carTable.removeChild($trClickedButton);
      },

      createCarTd: function createCarTd(input) {
        var $td = doc.createElement('td');
        $td.textContent = this.getCarInfo(input);
        return $td;
      },

      getCarInfo: function getCarInfo(input) {
        return $(input).get().value;
      },

      createCarImage: function createCarImage(carInfo) {
        var $tdImage = this.createTd();
        var $image = doc.createElement('img');
        $image.setAttribute('src', this.getCarInfo(carInfo));
        $tdImage.appendChild($image);
        return $tdImage;
      },

      createTd: function createTd() {
        return doc.createElement('td');
      }
    };
  })();

  app.init();

})(window, document, window.DOM);

```

**Comment by fdaciuk** [comment]
Pra reduzir código você precisa abstrair! Pega o que tá duplicado e tente normalizar os dados. Aí você cria uma função que recebe parâmetros e retorna esse resultado que era duplicado, mas com os dados passados por parâmetro.

Fazendo isso você consegue reduzir bastante coisa =)

**Comment by Roger-Melo** [comment]
Entendi, professor.  

Tentei fazer mais abstrações e cheguei ao resultado abaixo.  
Reduziu um pouco, mas ainda está bem grandinho. Vc pode me  
dar dicas de como refatorá-lo ainda mais? 

Obs.: a função `createCarTr`, por exemplo, está recebendo 5 parâmetros,  
isso pode ser ruim, não é? 

```javascript
(function(win, doc, $) {
  'use strict';

  var app = (function appController() {
    return {
      init: function init() {
        this.companyInfo();
        this.registerNewCar();
        this.getRegistredCar();
        this.initEvents();
      },

      registerNewCar: function registerNewCar() {
        var ajax = new XMLHttpRequest();
        ajax.open('POST', 'http://localhost:3000/car');
        ajax.setRequestHeader('Content-Type', 'application/x-www-form-urlencoded');
        ajax.send('image=http://wikiapk.com/wp-content/uploads/2017/06/73765_featured.jpg&brandModel=BMW&year=2020&plate=GIT9658&color=red');
        ajax.addEventListener('readystatechange', function() {
          if(ajax.readyState === 4)
            console.log(JSON.parse(ajax.responseText).message);
        }, false);
      },

      getRegistredCar: function getRegistredCar() {
        var ajax = new XMLHttpRequest();
        ajax.open('GET', 'http://localhost:3000/car');
        ajax.send();
        ajax.addEventListener('readystatechange', function() {
          if(ajax.readyState === 4) {
            var cars = JSON.parse(ajax.responseText);
            cars.forEach(app.getCarsFromServer);
          }
        }, false);
      },

      getCarsFromServer: function getCarsFromServer(car) {
        app.insertCarInTable(app.createCarTr(
          app.createCarImage(car.image),
          app.createCarTd(car.brandModel),
          app.createCarTd(car.year),
          app.createCarTd(car.plate),
          app.createCarTd(car.color)
        ));
      },

      createCarTr: function createCarTr(image, brandModel, year, plate, color) {
        var $tr = doc.createElement('tr');
        $tr.appendChild(image);
        $tr.appendChild(brandModel);
        $tr.appendChild(year);
        $tr.appendChild(plate);
        $tr.appendChild(color);
        $tr.appendChild(app.createRemoveButton());
        return $tr;
      },

      handleSubmit: function handleSubmit(e) {
        e.preventDefault();
        app.insertCarInTable(app.createCarTr(
          app.createCarImage(app.getInputValue('[data-js="image"]')),
          app.createCarTd(app.getInputValue('[data-js="brand-model"]')),
          app.createCarTd(app.getInputValue('[data-js="year"]')),
          app.createCarTd(app.getInputValue('[data-js="plate"]')),
          app.createCarTd(app.getInputValue('[data-js="color"]'))
        ));
      },

      insertCarInTable: function insertCarInTable(tr) {
        var $fragment = doc.createDocumentFragment();
        var $carTable = $('[data-js="car-table"]').get();
        $fragment.appendChild(tr);
        return $carTable.appendChild($fragment);
      },

      createCarTd: function createCarTd(tdContent) {
        var $td = doc.createElement('td');
        $td.textContent = tdContent;
        return $td;
      },

      createCarImage: function createCarImage(carImage) {
        var $tdImage = app.createTd();
        var $image = doc.createElement('img');
        $image.setAttribute('src', carImage);
        $tdImage.appendChild($image);
        return $tdImage;
      },

      getInputValue: function getInputValue(input) {
        return $(input).get().value;
      },

      createTd: function createTd() {
        return doc.createElement('td');
      },

      companyInfo: function companyInfo() {
        var ajax = new XMLHttpRequest();
        ajax.open('GET', '/company.json', true);
        ajax.send();
        ajax.addEventListener('readystatechange', app.getCompanyInfo, false);
      },

      getCompanyInfo: function getCompanyInfo() {
        if(!app.isRequestOk.call(this))
          return;
        var data = JSON.parse(this.responseText);
        app.fillCompanyInfo(data.name, data.phone);
      },

      isRequestOk: function isRequestOk() {
        return this.readyState === 4 && this.status === 200;
      },

      fillCompanyInfo: function fillCompanyInfo(dataName, dataPhone) {
        var $companyName = $('[data-js="company-name"]');
        var $companyPhone = $('[data-js="company-phone"]');
        $companyName.get().textContent = dataName;
        $companyPhone.get().textContent = dataPhone;
      },

      initEvents: function initEvents() {
        $('[data-js="form-register"]').on('submit', app.handleSubmit);
      },

      createRemoveButton: function createRemoveButton() {
        var $tdButton = app.createTd();
        var $button = doc.createElement('button');
        $button.innerHTML = 'Remover';
        $button.classList.add('remove-btn');
        $button.addEventListener('click', app.handleClickRemoveButton, false);
        $tdButton.appendChild($button);
        return $tdButton;
      },

      handleClickRemoveButton: function handleClickRemoveButton(e) {
        e.preventDefault();
        var $carTable = $('[data-js="car-table"]').get();
        var $trClickedButton = this.parentNode.parentNode;
        $carTable.removeChild($trClickedButton);
      }
    };
  })();

  app.init();

})(window, document, window.DOM);
```

**Comment by fdaciuk** [comment]
Legal @Roger-Melo! Vamos lá:
Realmente, receber muitos parâmetros em uma função não deixa o código muito legível. O ideal é sempre passar um parâmetro só, no máximo 2, que em alguns casos faz sentido.

Quando você precisa passar muitos parâmetros, recomendo usar um objeto, que assim você nomeia cada parâmetro, e não precisa se preocupar com a ordem, pois o objeto vai ser um parâmetro só que vai ser passado para a função, e as chaves dele não importam a ordem =)

Bom, pra fazer manipulação de DOM de forma manual, não dá pra fugir muito de código longo, mas tem algumas coisas que ainda você consegue abstrair:

A parte de requests, por exemplo: veja que o código é praticamente o mesmo. Você consegue abstrair isso em uma lib separada se quiser, assim como fizemos com a lib `DOM`. Aí você usa pode passar uma função de callback que vai ser executada no evento de `readystatechange`. Assim você manipula cada request de forma separada =)

No mais acho que é isso mesmo cara. Parabéns :D

**Comment by Roger-Melo** [comment]
Muito obrigado, professor. Vou tentar fazer essas abstrações e em breve lhe mostro o resultado!

**Comment by fdaciuk** [comment]
Show de bola! Qualquer dúvida, só avisar =)
